### PR TITLE
ADR-0004 - Out-of-the-box image repository in Stonesoup

### PR DESCRIPTION
**Pull request by sbose78**
Longer discussion https://docs.google.com/document/d/1_1xB6aBB7XW6JyYZVdnS8mC5Ftg4oksSThiLt8fBUcs/edit#

**Problem**
AppStudio does not have a internal registry where images could be pushed to as an intermediate step before being deployed as a container.
As an application centric experience, AppStudio should not make it mandatory for its users to specify where the image ( built from source code ) needs
to be pushed to.

**Goals**: 
* Provide an out-of-the-box location for users’ images to be pushed after being built from source.

### CODE REVIEW

**Review by alexeykazakov — APPROVED** [review]

File: ADR/0004-component-image-location.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+#### Short-term
+
+##### Setup
+* Configure the a pre-configured Quay.io API token for repo/org management, also known as the OAuth Access token. From a security perspective, this would be treated as a typical service account token
+  belonging to AppStudio. This token would be shortlived and rotated *as frequently as practically possible*  .
+
+##### When a user creates a Component
+1. Create a new Quay.io org owned by David using the pre-configured AppStudio-wide API token `quay.io/unique-org-david`
+2. Create a new image repo for Component “foo” using the the pre-configured David’s API token `quay.io/unique-org-david/unique-component-foo`
+3. Optionally, if needed, make the image repo public using the API token for easy image pulls.
```
Comment (gorkem) [inline ADR/0004-component-image-location.md:78]: We can not make the repository public. The user owns the repo, it should be their decision to make it public. Otherwise, we will risk exposing users' private applications. Other than David nobody else should be able to access it.

File: ADR/0004-component-image-location.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+4. Configure a robot account  `quay.io/unique-org-david` to be able to push to `quay.io/unique-org-david/unique-component-foo`
```
Comment (gorkem) [inline ADR/0004-component-image-location.md:79]: Is the token only for push? Or push and pull?

File: ADR/0004-component-image-location.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+* Security:
+  * Leverage short-lived Quay.io API tokens for repo/org management on Quay.io
+  * Leverage robot accounts for pushing images from CI/Build pipelines.
+
+
+
+## Decision
+
+### What
+
```
Comment (gorkem) [inline ADR/0004-component-image-location.md:51]: Can we also discuss here what happens to the repo/images when the workspace or user is deleted?

File: ADR/0004-component-image-location.md (hunk 2 above)
Comment (sbose78) [reply ADR/0004-component-image-location.md:79]: Push and Pull.

File: ADR/0004-component-image-location.md (hunk 1 above)
Comment (sbose78) [reply ADR/0004-component-image-location.md:78]: sounds good 👍 "private" is anyway the 'default' in quay.io . For private repositories, we'd have to propagate the pull secret to the deployment environment.

File: ADR/0004-component-image-location.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+* Per workspace, setup a new org “quay.io/unique-org-david/…”
```
Comment (ralphbean) [inline ADR/0004-component-image-location.md:52]:
```suggestion
* Per workspace called "david", setup a new org “quay.io/unique-org-david/…”
```

File: ADR/0004-component-image-location.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+* Create a new Quay.io org owned by David using David’s API token “quay.io/unique-org-david”
+* Create a new image repo for “foo” using David’s API token “quay.io/unique-org-david/appname-foo”
+* Optionally, if needed, make the image repo public using the David’s API token for easy image pulls.
+* Configure the robot account “appstudio” in “quay.io/unique-org-david” to be able to push to “quay.io/unique-org-david/appname-foo” Invalidate/Delete/Prune David’s API token. 
+
+
+#### Short-term
+
+##### Setup
+* Configure the a pre-configured Quay.io API token for repo/org management, also known as the OAuth Access token. From a security perspective, this would be treated as a typical service account token
```
Comment (ralphbean) [inline ADR/0004-component-image-location.md:72]:
```suggestion
* Configure a pre-configured Quay.io API token for repo/org management, also known as the OAuth Access token. From a security perspective, this would be treated as a typical service account token
```

File: ADR/0004-component-image-location.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  belonging to AppStudio. This token would be shortlived and rotated *as frequently as practically possible*  .
```
Comment (ralphbean) [inline ADR/0004-component-image-location.md:73]: Can we say which AppStudio service this belongs to? Does it belong specifically to `build-service`?

File: ADR/0004-component-image-location.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+We would use the user's API token for working with the relevant Quay.io organizations and repositories.
+
+
+## Open Questions
+
+- Since the long-term goal is to have the Quay.io organizations owned by the user, how do we build a frictionless experience to map the user's 
+  account with the user's Quay.io account ?
+- Considering the above is figured out, we would need to add existing users as owners to the relevant organizations. This is a backend job that 
+  would need to be designed and executed at an appropriate time.
+
```
Comment (ralphbean) [inline ADR/0004-component-image-location.md:94]:
```suggestion
- In the HACBS context, giving ownership of the image repo to users is not a problem for provenance, because integrity of images is based on attestations which are only treated as valid if signed by pipeline-service (by tekton chains). While users can in theory push images built on their laptop, the [release-service](../book/release-service.md) won't treat them as if they were built by the [pipeline-service](https://github.com/openshift-pipelines/pipeline-service).
```

File: ADR/0004-component-image-location.md (hunk 4 above)
Comment (ralphbean) [reply ADR/0004-component-image-location.md:52]: Do I understand it right?

File: ADR/0004-component-image-location.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+* Alexey Kazakov
+* Gorkem Ercan
+
+## Reviewers
+
+* Alexey Kazakov
+* Gorkem Ercan
+* Parag Dave
+
```
Comment (ralphbean) [inline ADR/0004-component-image-location.md:19]:
```suggestion
* Ralph Bean
```

File: ADR/0004-component-image-location.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (sbose78) [reply ADR/0004-component-image-location.md:52]: 👍 I wanted to convey that a unique org name is generated, I don't quite have an opinion on the generation function.

Comment (sbose78) [reply ADR/0004-component-image-location.md:52]: Correct.

File: ADR/0004-component-image-location.md (hunk 6 above)
Comment (sbose78) [reply ADR/0004-component-image-location.md:73]: I'm not really sure since build service is a mere consumer of this ?

File: ADR/0004-component-image-location.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+#### Short-term
+
+##### Setup
+* Configure the a pre-configured Quay.io API token for repo/org management, also known as the OAuth Access token. From a security perspective, this would be treated as a typical service account token
+  belonging to AppStudio. This token would be shortlived and rotated *as frequently as practically possible*  .
+
+##### When a user creates a Component
+1. Create a new Quay.io org owned by David using the pre-configured AppStudio-wide API token `quay.io/unique-org-david`
+2. Create a new image repo for Component “foo” using the the pre-configured David’s API token `quay.io/unique-org-david/unique-component-foo`
```
[the remark this answers is not in the document]
Comment (sbose78) [reply ADR/0004-component-image-location.md:76]: The UI should trigger the oauth flow with Quay.io using SPI to get us the API token.

File: ADR/0004-component-image-location.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+Date: Oct 29, 2022 
+
+## Status
+
+Accepted
+
+## Approvers
+
+* Alexey Kazakov
+* Gorkem Ercan
```
Comment (amfred) [inline ADR/0004-component-image-location.md:12]: I want to be an approver on this because it's a security requirement

**Review by amfred — CHANGES_REQUESTED** [review]
I think this proposed design will meet our isolation requirements, but there are a few open issues to resolve in the comments.

File: ADR/0004-component-image-location.md (hunk 10 above)
Comment (sbose78) [reply ADR/0004-component-image-location.md:12]: Sure @amfred

**Comment by amfred** [comment]
To me, the big open question is: which service is going to maintain the lifecycle of these Quay repos? We need to clean them up when we close accounts.

**Comment by sbose78** [comment]
@amfred , I'll update the ADR addressing that. We'll most likely have a delayed approach to deleting the Quay.io repositories to let users move their images to a different location in case they were depending on them for deployments.

**Comment by sbose78** [comment]
Update: In the initial phase, we'll have all repos under a single org. Once the concept of workspaces are better fleshed out and implemented, we'll move to one org per workspace. TODO: Confirm from the Quay.io team if programmatic creation of Quay.io orgs could be enabled.

**Comment by sbose78** [comment]
> TODO: Confirm from the Quay.io team if programmatic creation of Quay.io orgs could be enabled.

Confirmed, this is implementable. Either way, I've defined what our iterative approach towards implementing this should be.

> It's definitely one of our data retention policies. I put a proposal in our record retention policy (which we may want to publish here). My proposal is that we would disable accounts for 30 days after the subscription expires 

+1, I added that to the ADR as a "delayed" deletion.


> I think we should delete the images from our internal image registry at the same time as deleting the other resources (deployments, application configuration, etc.). 

+1 , added that to the ADR

> Some service is going to have to manage the lifecycle of this.

Yeah, the initial goal of this ADR wasn't to get into the details of which service would implement it - rather what would the service need to do. Over time, I can confirm that I have some clarity that this needs to be an independent controller given that pull secret is of interest for both build and deploy scenarios.

**Comment by arewm** [comment]
Asking about this after reading the comments but before reading the ADR itself.
>  We'll most likely have a delayed approach to deleting the Quay.io repositories to let users move their images to a different location in case they were depending on them for deployments.

Would any pull tokens generated (i.e. for private image repos) be deleted when the account/workspace is deleted? Would there be a way for a user to reactivate their account to reclaim the quay repos/organization if within the cleaning deadline?

File: ADR/0004-component-image-location.md [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+```
+    
+    
+
+
+## Open Questions
+
+- What would be a progammatic way to regenerate the main Quay.io API token ?
+- Since the long-term goal is to have the Quay.io organizations owned by the user, how do we build a frictionless experience to map the user's 
+  account with the user's Quay.io account ?
```
Comment (arewm) [inline ADR/0004-component-image-location.md:170]: Is there a way to transition between the types of quay ownership -- i.e. being able to hand over "keys" to an organization for an additional cost?

File: ADR/0004-component-image-location.md [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+- Considering the above is figured out, we would need to add existing users as members of the relevant organizations. This is a backend job that 
+  would need to be designed and executed at an appropriate time.
+- In the HACBS context, giving ownership of the image repo to users is not a problem for provenance, because integrity of images is based on attestations which are only treated as valid if signed by pipeline-service (by tekton chains). While users can in theory push images built on their laptop, the [release-service](../book/release-service.md) won't treat them as if they were built by the [pipeline-service](https://github.com/openshift-pipelines/pipeline-service). They will fail to pass the [enterprise contract](../book/enterprise-contract.md).
```
Comment (arewm) [inline ADR/0004-component-image-location.md:173]: Do we need to reword this as there isn't a really HACBS context anymore?

File: ADR/0004-component-image-location.md [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+2. Create an OAuth Application in the Quay.io organization.
+3. Geneate a token for the OAuth Application. This token would act as the 'service account' using which Quay.io resources would be created. Important to note, the token acts on behalf of the user who is requesting it - but uses the explicit scopes specified at the time of token generation. 
+4. Allowlist user 'shbose' to be create organizations using non-user-tokens using the Quay.io API.
+
+| Syntax      | Description |
+| ----------- | ----------- |
+| Quay.io organization      | quay.io/redhat-user-workloads       |
+| OAuth Application  name | Created, name redacted        |
+| Account used to generate token | `shbose` |
+| Scope | Administer organizations, Adminster repositories, Create Repositories |
```
Comment (arewm) [inline ADR/0004-component-image-location.md:73]: Can we align permissions granted to users for quay organizations/repositories to the Stonesoup roles as defined in https://github.com/redhat-appstudio/book/pull/42/files?

File: ADR/0004-component-image-location.md [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+* Token generation: 
+    * Robot account token: At the moment, the controller responsible for generating the Quay.io resources would be responsible for rotating the tokens https://docs.quay.io/api/swagger/#!/robot/regenerateUserRobotToken
+    * Quay.io API token: No programmatic way to regenerate this token is known at this point of time. This would be a manual activity to begin with. 
+
+* Upon deletion of an `Application`/`Component` from StoneSoup, 
+    * The controller/finalizer would delete the the relevant Quay.io resources namely, the image repository and the robot account.
+    * The controller/finalizer would delete the linked `Secret` from the user's namespace. Most likely, this should be a mere `ownerReference`-based garbage collection.
+
+* Upon removal of a user from Stonesoup,
+    * The empty Quay.io organization associated with the user or the user's Space may not be deleted instantly, but would be scheduled for a delayed cleanup.
```
[the remark this answers is not in the document]
Comment (sbose78) [reply ADR/0004-component-image-location.md:109]: >  Preventing others from being able to squat on it after cleanup?

Yes, this is what I had in mind.

[the remark this answers is not in the document]
Comment (arewm) [reply ADR/0004-component-image-location.md:109]: Gotcha, I wasn't sure if you had any other reason in mind. It doesn't seem like a necessary feature to protect against the quay repo itself (and potentially introduce unnecessary complexity). I imagine that it would be good to have a "delete process" where all the content which will be deleted is itemized so that the user can make an informed decision.

I also assume that any creation of organizations will be able to handle "fuzzing" to resolve name conflicts so there wouldn't be an issue if someone did manage to squat on an organization..

Is there any planned differentiation between empty and nonempty quay organizations?

File: ADR/0004-component-image-location.md [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+
+* Alexey Kazakov
+* Gorkem Ercan
+* Parag Dave
+* Ralph Bean
+
+## Context
+
+**Problem**
+StoneSoup does not have a internal registry where images could be pushed to as an intermediate step before being deployed as a container.
```
[the remark this answers is not in the document]
Comment (arewm) [reply ADR/0004-component-image-location.md:24]: Well, now change that to `AppStudio`. :)

File: ADR/0004-component-image-location.md (hunk 11 above)
Comment (sbose78) [reply ADR/0004-component-image-location.md:173]: That would be a customer support ticket I guess :-)

File: ADR/0004-component-image-location.md [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+* The quay.io repository created with use the format   quay.io/org-name/namespace-name/application-name/component-name. Example, https://quay.io/repository/redhat-user-workloads/shbose/applicationname/automation-repo
+
+#### Lifecycle of the Quay.io resources
+
+* Token generation: 
+    * Robot account token: At the moment, the controller responsible for generating the Quay.io resources would be responsible for rotating the tokens https://docs.quay.io/api/swagger/#!/robot/regenerateUserRobotToken
+    * Quay.io API token: No programmatic way to regenerate this token is known at this point of time. This would be a manual activity to begin with. 
+
+* Upon deletion of an `Application`/`Component` from StoneSoup, 
+    * The controller/finalizer would delete the the relevant Quay.io resources namely, the image repository and the robot account.
```
[the remark this answers is not in the document]
Comment (arewm) [reply ADR/0004-component-image-location.md:107]: Revisiting --

If there was a deletion delay to enable the repository to be re-hooked up to, would it even be _possible_ to create a new component which will push to the same location? Based on my current understanding, this isn't really possible without a new feature to "restore a recently deleted component." Therefore, this functionality would also be relegated to its own feature/request if desired.

> Could we support component archiving where artifacts remain but all configuration is unhooked?

Component images that have not been mirrored to some other location (i.e. via release pipelines) can be treated as "developer" images in that they are not guaranteed to remain for a long point in time. While these images can be treated a deletable (the behavior MUST be documented), an archival state seems like it could be useful as well -- but this would be a different feature/request.

> What if we have multiple components being pushed to the same repository but different versions (i.e. different branches)?

This still seems like a potentially valid use case to me. Do we afford users with enough customization and flexibility to push multiple components to one repository? I feel like we _should_ be able to afford this customization. Based on the current flow, I assume not ... so again, it would have to be a new feature request.

> I assume that this would have to vary if we are pushing to a custom quay organization.

Likely so, but this is far outside of the current support that it seems fine to defer this question until (and if) we decide to get there

File: ADR/0004-component-image-location.md (hunk 10 above)
Comment (Michkov) [reply ADR/0004-component-image-location.md:12]: @sbose78 are you going to add @amfred to fulfill this request?

File: ADR/0004-component-image-location.md [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+
+**Non-Goals**: 
+* Define the user experience for users bringing existing images from other image registry services.
+* Provide users an option to choose what the out-of-the-box location for images would be. 
+* Define the user experience for users who are willing to configure additional credentials for pushing to an image registry of their choice.
+
+**Design Goals**
+* Use Quay.io as the out-of-the-box image registry in order to avoid getting into the business of maintaining an internal registry.
+* Align user permissions in an StoneSoup workspace with those in Quay.io
+* Maintain the right levels of isolation between images being expected of a multi-tenant system.
```
Comment (tkdchen) [inline ADR/0004-component-image-location.md:43]: What are the levels of isolation? Is isolation level between images explained anywhere else in any AppStudio document?

File: ADR/0004-component-image-location.md [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+
+
+**Non-Goals**: 
+* Define the user experience for users bringing existing images from other image registry services.
+* Provide users an option to choose what the out-of-the-box location for images would be. 
+* Define the user experience for users who are willing to configure additional credentials for pushing to an image registry of their choice.
+
+**Design Goals**
+* Use Quay.io as the out-of-the-box image registry in order to avoid getting into the business of maintaining an internal registry.
+* Align user permissions in an StoneSoup workspace with those in Quay.io
```
Comment (tkdchen) [inline ADR/0004-component-image-location.md:42]: What does it mean by "those"?

File: ADR/0004-component-image-location.md [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+
+## Decision
+
+### What
+
+* Per workspace called "david", setup a new org “quay.io/unique-org-david/…”
+* Per component, setup a new new repo “quay.io/unique-org-david/unique-component-repo”
+* Use User’s Quay.io API token to manage the org/repo. Short-term, we'll use a pre-configured Quay.io API token associated with StoneSoup to create the org/repo till we 
+ figure out how to determinstically map a user in StoneSoup to a user in Quay.io.
+* Generate a robot account token scoped to the relevant repository and persist it in the user's workspace for the image build and push process to consume.
```
Comment (tkdchen) [inline ADR/0004-component-image-location.md:58]: Both "user's workspace" and "user's namespace" are used through this document. Do they mean the same thing? If yes, could we use single term consistently? If no, could we have a separate section to explain them as well as other terms used in this document?

File: ADR/0004-component-image-location.md [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+    * Robot account token: At the moment, the controller responsible for generating the Quay.io resources would be responsible for rotating the tokens https://docs.quay.io/api/swagger/#!/robot/regenerateUserRobotToken
+    * Quay.io API token: No programmatic way to regenerate this token is known at this point of time. This would be a manual activity to begin with. 
+
+* Upon deletion of an `Application`/`Component` from StoneSoup, 
+    * The controller/finalizer would delete the the relevant Quay.io resources namely, the image repository and the robot account.
+    * The controller/finalizer would delete the linked `Secret` from the user's namespace. Most likely, this should be a mere `ownerReference`-based garbage collection.
+
+* Upon removal of a user from Stonesoup,
+    * The empty Quay.io organization associated with the user or the user's Space may not be deleted instantly, but would be scheduled for a delayed cleanup.
+* PR-based tags are to be deleted on a regular basis. Image tags associated with `main` may remain un-pruned for now.
```
Comment (tkdchen) [inline ADR/0004-component-image-location.md:112]: Would it be good to replace `main` with "the merge target branch"?

**Comment by sbose78** [comment]
Took this, I will address the remaining comments as a follow-up 👍 (appreciate you taking out time to review)
